src/future/blocking.rs

```diff
@@ -0,0 +1,94 @@
+use Async;
+use future::Future;
+use executor::{self, NotifyHandle};
+use task_impl::ThreadNotify;
+
+/// Provides thread-blocking operations on a future.
+///
+/// See [`blocking`](fn.blocking.html) documentation for more details.
+#[derive(Debug)]
+#[must_use = "futures do nothing unless used"]
+pub struct Blocking<T> {
+    inner: executor::Spawn<T>,
+}
+
+/// Provides thread-blocking operations on a future.
+///
+/// `blocking` consumes ownership of `future`, returning a `Blocking` backed by
+/// the future. The `Blocking` value exposes thread-blocking operations that
+/// allow getting the realized value of the future. For example,
+/// `Blocking::wait` will block the current thread until the inner future has
+/// completed and return the completed value.
+///
+/// **These operations will block the current thread**. This means that they
+/// should not be called while in the context of a task executor as this will
+/// block the task executor's progress.
+pub fn blocking<T: Future>(future: T) -> Blocking<T> {
+    let inner = executor::spawn(future);
+    Blocking { inner: inner }
+}
+
+impl<T: Future> Blocking<T> {
+    /// Query the inner future to see if its value has become available.
+    ///
+    /// Unlike `Future::poll`, this function does **not** register interest if
+    /// the inner future is not in a ready state.
+    ///
+    /// This function will return immediately if the inner future is not ready.
+    pub fn try_take(&mut self) -> Option<Result<T::Item, T::Error>> {
+        match self.inner.poll_future_notify(&NotifyHandle::noop(), 0) {
+            Ok(Async::NotReady) => None,
+            Ok(Async::Ready(v)) => Some(Ok(v)),
+            Err(e) => Some(Err(e)),
+        }
+    }
+
+    /// Block the current thread until this future is resolved.
+    ///
+    /// This method will drive the inner future to completion via
+    /// `Future::poll`. **The current thread will be blocked** until the future
+    /// transitions to a ready state. Once the future is complete, the result of
+    /// this future is returned.
+    ///
+    /// > **Note:** This method is not appropriate to call on event loops or
```

Comment:
We should make the function *panic* in those cases, similarly to what was proposed in the RFC. Basically, just need to assert that there is not currently an executor. But maybe you're planning on doing that once we get the `Enter` stuff in?

Reply:
Yes, this step will happen in future PRs.